### PR TITLE
[WiP] FXP implementation

### PR DESCRIPTION
**Pull request by n0ix**
Hi (again),

i added [File eXchange Protocol (FXP)](https://en.wikipedia.org/wiki/File_eXchange_Protocol) Protocol to the FluentFTP libary.

**Why "Work in Progress" ?**
Cause there is a lot of code i added and i think there may be bugs or things didn't even think of. 
Also i think there is space for optimizations regarding reusing existing methods (cause some methods very similar).  I hope here you can help me, cause you know the whole code of fluent ftp much better then i do.

Also there is missing documentation of the new functions - i will do this later.

- n0ix

### CODE REVIEW

**Comment by robinrodricks** [comment]
Ohmigawd! This as awesome! A starting point is really good coz hey people can always contribute fixes as we go along. Its the initial setup that is the hardest to build, good job! I'll review it and let you know.

File: FluentFTP/Client/FtpClient_FXP.cs [hunk 1]
```diff
@@ -0,0 +1,1058 @@
...
+				var transferStarted = DateTime.Now;
+				long lastSize = 0;
+
+				var sourceFXPTransferReply = ftpFxpSession.sourceFtpClient.GetReply();
+				var destinationFXPTransferReply = ftpFxpSession.destinationFtpClient.GetReply();
+
+				while (!sourceFXPTransferReply.Success || !destinationFXPTransferReply.Success)
+				{
+
+					if (remoteClient.EnableThreadSafeDataConnections)
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_FXP.cs]: Why is it required to check EnableThreadSafeDataConnections here?

File: FluentFTP/Client/FtpClient_FXP.cs [hunk 2]
```diff
@@ -0,0 +1,1058 @@
...
+					{
+						FtpTrace.Write(FtpTraceLevel.Info, "reporting progress");
+						// send progress reports
+						if (progress != null && fileSize != -1)
+						{
+							offset = remoteClient.GetFileSize(remotePath);
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_FXP.cs]: Here you are getting the file size from the remote server how often?

File: FluentFTP/Client/FtpClient_FXP.cs [hunk 3]
```diff
@@ -0,0 +1,1058 @@
...
+
+			// get all the files in the local directory
+			var fileListing = GetListing(sourceFolder, FtpListOption.Recursive).Where(x => x.Type == FtpFileSystemObjectType.File).Select(x => x.FullName).ToArray();
+
+			// loop thru each file and transfer it
+			var filesToUpload = GetFilesToTransfer(sourceFolder, remoteFolder, rules, results, shouldExist, fileListing);
+			TransferServerFiles(filesToUpload, remoteClient, existsMode, verifyOptions, progress, remoteListing);
+
+			// delete the extra remote files if in mirror mode and the directory was pre-existing
+			// DeleteExtraServerFiles(mode, shouldExist, remoteListing);
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_FXP.cs]: A bit of code duplication here, but still great considering most of the inner functions are reused.

File: FluentFTP/Client/FtpClient_FXP.cs [hunk 4]
```diff
@@ -0,0 +1,1058 @@
...
+
+				// record that this file should exist
+				shouldExist.Add(remoteFile.ToLowerInvariant(), true);
+
+				// absorb errors
+				filesToTransfer.Add(result);
+			}
+
+			return filesToTransfer;
+		}
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_FXP.cs]: Isn't this total code duplication? Is there a way to make a common method so the Upload/Download Directory methods can reuse this with FXP? or is it too many changes?

File: FluentFTP/Client/FtpClient_FXP.cs [hunk 5]
```diff
@@ -0,0 +1,1058 @@
...
+			}
+
+			if (!this.IsConnected)
+			{
+				throw new FluentFTP.FtpException("The source FXP FtpClient must be open and connected before a transfer between servers can be intitiated");
+			}
+
+			if (!await FileExistsAsync(sourcePath, token)){
+				throw new FluentFTP.FtpException(string.Format("Source File {0} cannot be found or does not exists!", sourcePath));
+			}
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_FXP.cs]: This is a good, safe check, but is it needed again at this point?

File: FluentFTP/Client/FtpClient_Properties.cs [hunk 6]
```diff
@@ -795,6 +795,17 @@ public partial class FtpClient : IDisposable {
 			set => m_DownloadDataType = value;
 		}
 
+		public FtpDataType m_FXPDataType = FtpDataType.Binary;
+
+		/// <summary>
+		/// Controls if the high-level API transfer (FXP) files in Binary or ASCII mode.
```
Comment (robinrodricks) [inline FluentFTP/Client/FtpClient_Properties.cs]: Change comment:

Controls if the FXP server-to-server file transfer API uses Binary or ASCII mode.

File: FluentFTP/Client/FtpClient_FXP.cs (hunk 1 above)
Comment (n0ix) [reply FluentFTP/Client/FtpClient_FXP.cs]: Cause we can't do this when we have only one connection, the one which is open is already used for the FXP Transfer and can't answer any other commands. Maybe i got this wrong, but a ftp client should/can be onlycloned if ThreadSafeDataConnections are enabled. In OpenPassiveFXPConnection i clone the connection if ThreadSafeDataConnections is enabled so we can later track progress.

File: FluentFTP/Client/FtpClient_FXP.cs (hunk 2 above)
Comment (n0ix) [reply FluentFTP/Client/FtpClient_FXP.cs]: Very often, and i know this is ugly - but i did find any other possibility to track progress of the FXP Transfer (cause we have no stream here). May be we can add another Property which enables this?

File: FluentFTP/Client/FtpClient_FXP.cs (hunk 3 above)
Comment (n0ix) [reply FluentFTP/Client/FtpClient_FXP.cs]: Yes i agree here is definitly space to deduplicate code. The problem are the "directions" in FXP. The Upload und Download methods i can't reuse cause the assume there is a a local path which we don't have when we use fxp.

File: FluentFTP/Client/FtpClient_FXP.cs (hunk 4 above)
Comment (n0ix) [reply FluentFTP/Client/FtpClient_FXP.cs]: as already mentioned before, reuse of this methods is difficult cause the assume a local path (as source or as destination) which don't have here.

File: FluentFTP/Client/FtpClient_FXP.cs (hunk 5 above)
Comment (n0ix) [reply FluentFTP/Client/FtpClient_FXP.cs]: Yeah cause it's never have been checked before or i'am wrong?

**Comment by robinrodricks** [comment]
Thank you for the effort!

**Comment by robinrodricks** [comment]
Thanks for this fantastic contribution. I've cleaned up the code to remove duplicate sections and also improved the comments. I noticed how you stayed within the guidelines by using the FolderUpload/FolderDownload logic as a template. I also fixed some logging (LogFunc) and code logic. 

New methods that are reused between Folder Upload and Folder Transfer (FXP):

- `CanUploadFile`
- `CreateSubDirectories`

New methods that are reused in File Transfer (FXP):

- `VerifyTransferFileParams`

**Comment by robinrodricks** [comment]
Released in 31.3.0!


  - New: `TransferFile` and `TransferDirectory` methods to transfer files from server to server (thanks [n0ix](/n0ix))
  - New: FXP (File Transfer Protocol) implementation to support direct server-to-server transfers (thanks [n0ix](/n0ix))

**Comment by robinrodricks** [comment]
Added an icon and a note in the top features on the [readme](https://github.com/robinrodricks/FluentFTP). Will do documentation soon.

**Comment by n0ix** [comment]
Awesome! Thanks for the code cleanup an the quick release!

**Comment by robinrodricks** [comment]
Welcome. When is the FtpFxpSession destroyed? We might need to destroy the cloned connection to the remote server after transfer.

**Comment by robinrodricks** [comment]
Documented! - https://github.com/robinrodricks/FluentFTP/wiki/FXP-Server-To-Server

**Comment by n0ix** [comment]
> 
> 
> Welcome. When is the FtpFxpSession destroyed? We might need to destroy the cloned connection to the remote server after transfer. We probably don't need to destroy the Source server connection, am I right? Because if that is the main connection then why should we destroy it in `FtpFxpSession.Dispose`?

Hm, yeah we should clean up this, but only if we **cloned** the connections. Else we need to leave as them are cause the user created them manually and won't expect that one the connections getting disposed. Is there an property on cloned connections so we can identify them? Then it should be easy to dispose the `FxpSession `in the finally block of `TransferFile `/ `TransferDirectory`

**Comment by robinrodricks** [comment]
I've tried to handle this, can you check the latest trunk and let me know if it works?

Also, can we reuse the same FXP session for an entire folder instead of creating one per file transfer? I'll do the refactoring, just tell me if its possible and recommended.

**Comment by robinrodricks** [comment]
Newly released in 32.2.0:

  - New: `GetChecksum` allows you to specify a hash algorithm to be run on the server if supported
  - New: FXP file transfer now validates the file using the first mutually supported algorithm

Can you check if the FXP logic for checking mutually supported algorithm is correct? It is based on the new feature in GetChecksum that allows you to execute a specific hash on the server rather than executing any random hash.

**Comment by robinrodricks** [comment]
Please see - https://github.com/robinrodricks/FluentFTP/issues/546

**Pull request merged**